include/circt/Support/Backedge.h

```diff
@@ -0,0 +1,76 @@
+//===- Backedge.h - Support classes for building backedges ------*- C++ -*-===//
+//
+// Backedges are temporary operations/values which have to exist as operands
+// before they are produced in a result. Since it isn't clear how to build
+// backedges in MLIR, these helper classes set up a canonical way to do so.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_SUPPORT_BACKEDGE_H
+#define CIRCT_SUPPORT_BACKEDGE_H
+
+#include "mlir/IR/Location.h"
+#include "mlir/IR/Value.h"
+#include "llvm/ADT/SmallVector.h"
+
+namespace mlir {
+class PatternRewriter;
+class Operation;
+} // namespace mlir
+
+namespace circt {
+
+class Backedge;
+
+/// Instantiate one of these and use it to build typed backedges. Backedges
+/// which get used as operands must be assigned to with the actual value before
+/// this class is destructed, usually at the end of a scope. It will check that
+/// invariant then erase all the backedge ops during destruction.
+///
+/// Example use:
+/// ```
+///   circt::BackedgeBuilder back(rewriter, loc);
+///   circt::Backedge ready = back(rewriter.getI1Type());
+///   // Use `ready` as a `Value`.
+///   auto addOp = rewriter.create<addOp>(loc, ready);
+///   // When the actual value is available,
+///   ready = anotherOp.getResult(0);
+/// ```
+class BackedgeBuilder {
+  friend class Backedge;
+
+public:
+  /// To build a backedge op and manipulate it, we need a `PatternRewriter` and
+  /// a `Location`. Store them during construct of this instance and use them
+  /// when building.
+  BackedgeBuilder(mlir::PatternRewriter &rewriter, mlir::Location loc);
```

Comment:
Does this actually need a PatternRewriter, or is an OpBuilder enough?

Reply:
Yeah, it does. OpBuilder doesn't have .eraseOp() which is what is supposed to be used when deleting an op in a pattern conversion, which is the initial use-case. We can likely evolve this to include more use-cases.